Support integration branch bisection for Fenix (Bug 1857026)

moztcampbell opened the pull request:
This improves support of Fenix and Focus by supporting bisection without the presence of build-info txt files on archive.mozilla.org. Instead of the text files, I use the existing JsonPushes modules to determine changeset based on build timestamp extracted from path names on archive.mozilla.org. With some small fixes to the FenixConfig, it is easy to support the Integration mixin and get proper pushlog support.

The current fenix builds use taskcluster routes called 'nightly' or 'nightly-simulation' that are roughly what we'd consider to be 'shippable' builds on desktop.

moztcampbell committed:
Bug 1857026 - Fixup Fenix/Focus nightly repo handling

The "nightly_repo" for these should be the VCS target which is mozilla-central,
while the "nightly_base_repo_name" can name used by archive.mozilla.org. Also
remove the arch_regex_bits variable which isn't helping.

moztcampbell committed:
Bug 1857026 - Use json-pushes API to get build info for Fenix

The Fenix nightlies on archive.mozilla.org don't contain the usual build info
txt files. As an alternative, use the hg.mozilla.org json-pushes API to look
up based on buildid (datetime).

Add a new interface to JsonPushes to look up the push at a specific timestamp
rather than just a range of dates.

moztcampbell committed:
Bug 1857026 - Add integration branch bisection for Fenix/Focus

Set the name of the build type to "shippable" since the builds are all PGO. Note
that the task names for Fenix call themselved 'nightly-simulation' but the intent
is the same as 'shippable'.

This also adds support for Focus, but that is less valuable since we don't
schedule compatible builds very often.

moztcampbell commented on mozregression/fetch_configs.py:
```diff
@@ -425,23 +433,22 @@ def get_nightly_repo_regex(self, date):
 
     def _get_nightly_repo(self, date):
-        return "fenix"
+        return "mozilla-central"
```
This needs to be the VCS repo, which for Fenix is now mozilla-central.

zzzeid replied:
Will this impact historical builds, before the move to mozilla-central? We may have to add some fallback mechanisms.

moztcampbell replied:
Old Code:
 - `nightly_base_repo_name` used to identify taskcluster routes and archive.m.o paths
 - `get_nightly_repo_regex` used `get_nightly_repo` incorrectly I'd argue
 - The `get_nightly_repo` used for desktop builds integration branches
 - Fenix doesn't support integration branch bisection

New Code:
 - `get_nightly_repo_regex` now uses `nightly_base_repo_name` since that still says "fenix" and still is used for archive.m.o paths
 - Both fenix and desktop use `get_nightly_repo` to find the source branch

So, the fenix artifacts from `fenix` repo to `mozilla-android` repo to `mozilla-central` repo all published to same taskcluster routes and same naming scheme for archive.m.o. Good question though since this part drove me a little mad to figure out what is breaking for me.

moztcampbell replied:
Old fenix artifacts will still be fetchable from pre-monorepo with this code since they come from archive.m.o. Any integration branch artifacts are long gone from task cluster. We never published any integration artifacts on archive.m.o as we do for Desktop (eg https://archive.mozilla.org/pub/firefox/integration/autoland/2024/12/06/09af335a2b3b0e2ccf214720645de29b51407509/) so Fenix can't do fine-grained bisection beyond the time range of taskcluster artifacts (like 3 months or so).

moztcampbell replied:
I'll add the other repo references and comments though

moztcampbell committed:
Bug 1857026 - Use first BUILD_TYPE as default

To allow configs that don't target "opt", use the first build type in the
list as the default. Correct some minor test expectations.

zzzeid commented on mozregression/fetch_build_info.py:
```diff
@@ -212,6 +212,21 @@ def _fetch_build_info_from_url(self, url, index, lst):
+            infos["repository"] = jpushes.repo_url
+            infos["changeset"] = push.changeset
+        except Exception:
+            LOG.info(f"Unable to fetch push by timestamp for {build_dt}")
```
Ideally this try/except block would be a lot more specific. We can probably modify it like this:

```suggestion
        try:
            branch = self.fetch_config.get_nightly_repo(date)
        except <a more specific exception, if we know it>:
            LOG.info(f"Unable to fetch push by timestamp for {build_dt}")
        else:
            jpushes = JsonPushes(branch=branch)
            push = jpushes.push_by_timestamp(build_dt)[-1]
            infos["repository"] = jpushes.repo_url
            infos["changeset"] = push.changeset
```

But I think at least we can probably move some of the items to the else block.

zzzeid commented on mozregression/fetch_configs.py:
```diff
@@ -425,23 +433,22 @@ def get_nightly_repo_regex(self, date):
 
 class FenixNightlyConfigMixin(NightlyConfigMixin):
     nightly_base_repo_name = "fenix"
-    arch_regex_bits = ""
+    has_build_info = False
+    has_json_pushes = True
 
     def _get_nightly_repo(self, date):
-        return "fenix"
+        return "mozilla-central"
 
     def get_nightly_repo_regex(self, date):
-        repo = self.get_nightly_repo(date)
-        repo += self.arch_regex_bits  # e.g., ".*arm64.*".
+        repo = f"{self.nightly_base_repo_name}-[^-]+-android"
+        if self.arch:
+            repo += f"-{self.arch}"
```
Same question for these lines, does this need to accommodate different repo regexes based on the date?

moztcampbell replied:
Similarly, the scheme for archive.m.o was consistent across all repo migrations. My "fix" is to use 'nightly_base_repo_name' (which is the taskcluster/archive.m.o name) instead of 'get_nightly_repo' which is the active taskcluster repo.

zzzeid commented on mozregression/fetch_build_info.py:
```diff
@@ -277,6 +292,8 @@ def find_build_info(self, date, fetch_txt_info=True, max_workers=2):
                 if fetch_txt_info:
                     self._update_build_info_from_txt(infos)
+                if ("changeset" not in infos) and self.fetch_config.has_json_pushes:
+                    self._update_build_info_from_json_pushes(date, infos)
```
I think this is getting shoe-horned into the NightlyInfoFetcher when it should probably fall back to IntegrationInfoFetcher (e.g., via IntegrationHandler). I.e., we need a FenixIntegrationConfigMixin in fetch_configs.

moztcampbell commented:
I think you are right that I have misplaced the patch that has `FenixIntegrationConfigMixin`....

moztcampbell committed:
Bug 1857026 - Refactor NightlyInfoFetcher with get_nightly_changeset

Add `ArchiveBuildUrls` and `ChangesetInfo` types to refactor the
`NightlyInfoFetcher` so that the config can control how changeset is resolved
instead of only using the .txt file mechanism. The config now provides a
`get_nightly_changeset` method that examines build URLs scraped from server and
then computes the corresponding changeset/repo if possible.

The `has_build_info` property is also removed and instead we rely on a
`get_nightly_info_url` method providing a non-default value.

moztcampbell committed:
Bug 1857026 - Fix thunderbird-l10n nightly pushlogs

The `has_build_info` property is set to False, but a `get_nightly_info_url`
was not provided. Fix this and add a test case.

moztcampbell commented:
In the second patch in stack, I introduce `FetchInfo` types as helpers to the NightlyInfoFetcher and allow the fetch_configs to control which variant to use. This clarifies the fetch_configs that don't use an `changeset` resolution for nightlies, and the later patches in stack introduce the pushlog-based version.
I still maintain the call to json-pushes distinct from the IntegrationHandler since resolving in NightlyHandler means I get push range urls while bisecting the nightlies before the IntegrationHandler actually kicks off.
Feedback welcome on the fetch_configs interfaces. I have them return a class to control the FetchInfo type, but maybe there are better approaches.

moztcampbell commented:
Here is my complete stack with more testing and more bug fixes.

- I've broken this down into a stack of patches that each pass tests
- This has the FetchInfo/TxtFetchInfo/PushlogFetchInfo refactor so that a fetch config can specify which to use
- This added a few more tests for the NightlyInfoFetcher for the fenix case
- I changed how mixins are inherited to be at start of list since I was hitting MRO issues trying to override methods and these seems consistent with Python recommendations as best I can tell
- The FenixIntegrationConfig is now included ends up being simplified
- I've added more internal tests and done more manual testing of this version
- I ensured that even without integration-bisection that the nightly bisection is able to generate the correct pushlogs

zzzeid commented on mozregression/fetch_build_info.py:
```diff
@@ -172,9 +142,88 @@ def find_build_info(self, push):
+class FetchInfo:
+    """
+    Holds data about a candidate builds fetched from the nightly archive
+    servers.
```
Summary should fit on the first line and respect max width (see [pep-0257](https://peps.python.org/pep-0257/#multi-line-docstrings) for more detail). More elaborate detail can go after the blank line, if needed.

This applies to any other docstrings in this PR that are similar.

zzzeid commented on mozregression/fetch_build_info.py:
```diff
@@ -172,9 +142,88 @@ def find_build_info(self, push):
+            and self.changeset == other.changeset
+            and self.repository == other.repository
+        )
+
```
It looks like this is added just for usage in tests. It would be better to remove this and instead either add a method that is used in tests, to do what this is doing, or to simply check these things explicitly in the tests.

zzzeid commented on mozregression/fetch_build_info.py:
```diff
@@ -172,9 +142,88 @@ def find_build_info(self, push):
+            and self.repository == other.repository
+        )
+
+    def update_metadata(self, fetch_config):
```
nit:
```suggestion
    def update_info(self, fetch_config):
```

zzzeid commented on mozregression/fetch_build_info.py:
```diff
@@ -183,34 +232,41 @@ def _fetch_build_info_from_url(self, url, index, lst):
+            build_txt = build_txt_urls[0] if build_txt_urls else None
+
+            fetch_info_class = self.fetch_config.get_nightly_fetch_info_class()
+            fetch_info = fetch_info_class(build, build_txt)
```
```suggestion
            fetch_info = self.fetch_config.fetch_info(build, build_txt)
```

zzzeid commented on mozregression/fetch_configs.py:
```diff
@@ -262,6 +263,13 @@ def extra_persist_part(self):
         """
         return ""
 
+    def get_nightly_fetch_info_class(self):
```
```suggestion
    @property
    def info_class(self):
```

zzzeid commented on mozregression/fetch_configs.py:
```diff
@@ -424,23 +435,50 @@ def get_nightly_repo_regex(self, date):
         return self._get_nightly_repo_regex(date, repo)
 
+    def get_nightly_fetch_info_class(self):
+        return PushlogFetchInfo
```
I think this would improve readability

```suggestion
    @property
    def info_class(self):
        return PushlogFetchInfo
```

zzzeid commented on mozregression/fetch_configs.py:
```diff
@@ -684,7 +735,7 @@ class ThunderbirdL10nConfig(L10nMixin, ThunderbirdL10nNightlyConfigMixin, Common
 
 @REGISTRY.register("fennec")
-class FennecConfig(CommonConfig, FennecNightlyConfigMixin, FennecIntegrationConfigMixin):
+class FennecConfig(FennecNightlyConfigMixin, FennecIntegrationConfigMixin, CommonConfig):
```
I think this change here points to a slight problem in the implementation. Ideally we wouldn't need to "override" the `get_nightly_fetch_info_class` method. Instead, it should be provided by the correct mixin as needed. So if we want to keep this intended behaviour, I would detach `get_nightly_fetch_info_class` altogether from `CommonConfig` and it to the corresponding/necessary mixins instead. That would eliminate the need to change the order of classes here (and elsewhere.)

zzzeid commented on mozregression/json_pushes.py:
```diff
@@ -155,3 +155,23 @@ def push(self, changeset, **kwargs):
+        a MozRegressionError.
+        """
+
+        assert isinstance(timestamp, datetime.datetime)
```
Use of `assert` should be avoided outside of tests.
```suggestion
        if not isinstance(timestamp, datetime.datetime):
            raise TypeError()
```

zzzeid commented on mozregression/fetch_configs.py:
```diff
@@ -130,7 +131,7 @@ def __init__(self, os, bits, processor, arch):
         self.set_arch(arch)
         self.repo = None
-        self.set_build_type("opt")
+        self.set_build_type(self.BUILD_TYPES[0])
```
Nice. I wonder why this wasn't set like this originally.

zzzeid commented on mozregression/fetch_build_info.py:
```diff
@@ -172,9 +142,88 @@ def find_build_info(self, push):
         )
 
 
+class FetchInfo:
```
This is a nit, and I don't have a better suggestion off the top of my head, but I think the naming here might get a little confusing since we now have `InfoFetcher` and `FetchInfo`, and the main point of having this class is to have broader sources to fetch from. So maybe we should figure out a better separation here between the source of the info, and the info itself.

moztcampbell commented:
I did another pass at this and replaced `FetchInfo` I was adding with `ArchiveBuildUrls` and `ChangesetInfo` to reflect the two stages they are generated. This also made it easier to avoid passing a class around and instead using a mixin on the fetch config to control how the ChangesetInfo is computed.
In the process of doing this I discovered a new bug in the thunderbird-l10n builds where they weren't finding the text files even though they were partially configured for them.
